src/utils/cpg.ts

```diff
@@ -0,0 +1,258 @@
+import { Action, Pathway } from 'pathways-model';
+import {
+  ActivityDefinition,
+  PlanDefinition,
+  PlanDefinitionAction,
+  Bundle,
+  BundleEntry,
+  Library
+} from 'fhir-objects';
+import { v4 as uuidv4 } from 'uuid';
+import { isActionNode, findParents, isBranchNode } from './nodeUtils';
+import { R4 } from '@ahryman40k/ts-fhir-types';
+
+const LIBRARY_DRAFT = R4.LibraryStatusKind._draft;
+const PLANDEFINITION_DRAFT = R4.PlanDefinitionStatusKind._draft;
+const ACTIVITYDEFINITION_DRAFT = R4.ActivityDefinitionStatusKind._draft;
+const BUNDLE_TRANSACTION = R4.BundleTypeKind._transaction;
+const BUNDLE_PUT = R4.Bundle_RequestMethodKind._put;
+const CONDITION_APPLICABILITY = R4.PlanDefinition_ConditionKindKind._applicability; // eslint-disable-line
+const EXPRESSION_CQL = R4.ExpressionLanguageKind._textCql;
+
+export function createActivityDefinition(action: Action): ActivityDefinition {
+  const activityId = uuidv4();
+  const kind =
+    action.resource.resourceType === 'Procedure' ? 'ServiceRequest' : action.resource.resourceType;
+
+  const activityDefinition: ActivityDefinition = {
+    id: activityId,
+    resourceType: 'ActivityDefinition',
+    meta: {
+      profile: ['http://hl7.org/fhir/uv/cpg/StructureDefinition/cpg-publishableactivity']
+    },
+    extension: [
+      {
+        url: 'http://hl7.org/fhir/uv/cpg/StructureDefinition/cpg-knowledgeCapability',
+        valueCode: 'publishable'
+      }
+    ],
+    url: `urn:uuid:${activityId}`,
+    version: '1.0',
+    name: `AD${activityId.substring(0, 7)}`,
+    title: `ActivityDefinition: ${activityId}`,
+    status: ACTIVITYDEFINITION_DRAFT,
+    experimental: true,
+    date: new Date().toISOString(),
+    publisher: 'Logged in user',
+    description: action.description,
+    kind: kind,
+    productCodeableConcept:
+      action.resource.resourceType === 'MedicationRequest'
+        ? action.resource.medicationCodeableConcept
+        : action.resource.code
+  };
+
+  return activityDefinition;
+}
+
+function createAction(
+  id: string,
+  description: string,
+  definition: string,
+  hasConditions = false
+): PlanDefinitionAction {
+  const cpgAction: PlanDefinitionAction = {
+    id: id,
+    title: `Action: ${id}`,
+    description: description,
+    code: [
+      {
+        coding: [
+          {
+            system: 'http://hl7.org/fhir/uv/cpg/CodeSystem/cpg-common-process',
+            code: 'guideline-based-care',
+            display: 'Guideline-based Care'
+          }
+        ]
+      }
+    ],
+    definitionCanonical: definition
+  };
+  if (hasConditions) cpgAction.condition = [];
```

Comment:
Is it expected that in the future this will be populated or is it populated somewhere else?  Ultimately we wouldn't want to leave this empty because in FHIR JSON arrays are never empty.

Reply:
Its set in the `toCPG` function but at the time we create the action we don't know what the conditions are so we iterate over the parents and then set the condition

Reply:
since the `hasConditions` argument is only used to initialize this to an empty array and this method is only used in `toCPG` I think it makes sense to remove this argument and just handle setting the conditions in `toCPG` where it is known what the conditions are and that they exist.  This way we know that the array will always be populated if present.

i.e. on line 237

`cpgStrategyAction.condition?.push(condition);` can become something like:

```typescript
cpgStrategyAction.condition = cpgStrategyAction.condition || []
cpgStrategyAction.condition.push(condition);
```